File: packages/sdk-node/src/contract/loader.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { parse } from "yaml";
import { ContractSchema, type Contract } from "./schema.js";

export class ContractError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ContractError";
  }
}

export function loadContract(path: string): Contract {
  const abs = resolve(path);
  let raw: string;
  try {
    raw = readFileSync(abs, "utf-8");
  } catch {
    throw new ContractError(`Contract file not found: ${abs}`);
  }

  let data: unknown;
  try {
    data = parse(raw);
  } catch (e) {
    throw new ContractError(
      `Invalid YAML in ${abs}: ${e instanceof Error ? e.message : String(e)}`,
    );
  }

  if (typeof data !== "object" || data === null || Array.isArray(data)) {
    throw new ContractError(`Contract root must be a mapping, got ${typeof data}`);
  }

  const parsed = ContractSchema.safeParse(data);
  if (!parsed.success) {
    throw new ContractError(
      `Contract schema validation failed for ${abs}:\n${parsed.error.message}`,
    );
  }
  return parsed.data;
}

File: packages/core/src/mymcp/contract/loader.py
"""Load and validate contract YAML files."""

from __future__ import annotations

from pathlib import Path

import yaml
from pydantic import ValidationError

from .models import Contract


class ContractError(Exception):
    """Raised when a contract file is invalid or cannot be loaded."""


def load_contract(path: str | Path) -> Contract:
    p = Path(path)
    if not p.is_file():
        raise ContractError(f"Contract file not found: {p}")
    try:
        raw = yaml.safe_load(p.read_text(encoding="utf-8"))
    except yaml.YAMLError as e:
        raise ContractError(f"Invalid YAML in {p}: {e}") from e
    if not isinstance(raw, dict):
        raise ContractError(f"Contract root must be a mapping, got {type(raw).__name__}")
    try:
        return Contract.model_validate(raw)
    except ValidationError as e:
        raise ContractError(f"Contract schema validation failed for {p}:\n{e}") from e
